### ISSUE TITLE
constrained_layout and texts outside of the axes

### ISSUE THREAD

**Issue by anntzer** [matplotlib/matplotlib#19358]
### Bug report

**Bug summary**

Texts outside of the axes extents interact "interestingly" with constrained_layout.

**Code for reproduction**

```python
ax = figure(constrained_layout=True).add_subplot(); ax.plot([0, 1]); ax.text(2, 2, "hello")
ax = figure(constrained_layout=True).add_subplot(); ax.plot([0, 1]); ax.text(4, 4, "hello")
```

**Actual outcome**

The first example gives
![1](https://user-images.githubusercontent.com/1322974/105726883-cf881800-5f2a-11eb-9b84-6e6070a6f7b1.png)
the second one
![2](https://user-images.githubusercontent.com/1322974/105726903-d31b9f00-5f2a-11eb-8961-fed83e53bff9.png)
together with (only for the second one)
```
UserWarning: constrained_layout not applied because axes sizes collapsed to zero.  Try making figure larger or axes decorations smaller.
```

**Expected outcome**

I'm not actually sure how I'd actually want constrained_layout to handle such cases, but the current state seems... shaky.

**Matplotlib version**

  * Operating system: linux
  * Matplotlib version (`import matplotlib; print(matplotlib.__version__)`): HEAD
  * Matplotlib backend (`print(matplotlib.get_backend())`): qt5agg
  * Python version: 39
  * Jupyter version (if applicable): 
  * Other libraries:

**Comment by jklymak** [comment]
You can take the text out of the layout.  But I'm confused - is the text not clipped to the axes?  If it is clipped I guess it should not get a window_extent at all and hence should not affect the layout managers

**Comment by jklymak** [comment]
Then this is acting as expected?    Constrained/tight_layout says adjust the axes size so that all artists can be seen.  I agree that is sometimes not desirable, but sometimes it is, hence we have `clip_on` and `set_in_layout`.

**Comment by anntzer** [comment]
But here the text is not actually visible in the end layout in either example, even though constrained_layout *could* have resized the axes to achieve that.

**Comment by jklymak** [comment]
I mean theoretically, yes, but practically, its not surprising.  The layout managers make margins in figure space to accommodate artists placed in axes space.  If you place the artist in data space, it makes a margin assuming the artist will be at the same location in axes space, but the artist of course moves because it is being placed in data space and the axes has changed size.  This usually isn't a problem for clipped artists, but I am not surprised its a problem here.  I'm not sure there is anything to be done about it.

**Comment by jklymak** [comment]
... or maybe more succinctly: if the artist is placed in data units then the size of the margin required to accommodate the artist depends on the size of the margin, so you need to solve an inverse problem.  

Perhaps we should change the default of `in_layout` to depend on the transform of the object, and turn it off by default if the object is in data units?

**Comment by anntzer** [comment]
> Perhaps we should change the default of in_layout to depend on the transform of the object, and turn it off by default if the object is in data units?

This seems reasonable to me.

**Comment by jklymak** [comment]
I don't think this is a regression, however, so I don't think we need to rush a fix out the door.  Setting `in_layout` in a way that checks the transform and respects the user's setting it will require a bit of finagling.

**Comment by anntzer** [comment]
(I agree there's no hurry.)

**Comment by jklymak** [comment]
Actually, thinking about this some more, I'm not really sure what we can do.  The original motivation for making most artists in the layout were artists that were partially overhanging an axes and we wanted to make sure they don't hit the next axes.  Maybe that was a mistake, in that unclipped artists can be dragged out of view and then CL (and tight_layout) try to make a big margin to stop them from overlapping other (potential) axes.  

So we can either have a bunch of confusing heuristics for whether an artist is included in the layout or not, or we can just set to a default and trust the user to take things out if they are not working.  I'm not against the default being different for different artists.  i.e. I'm almost convinced that `text` artists should not be in the layout by default, and only included in special cases.  Certainly labelling a random data point should be allowed to scroll out of view.  OTOH, you could also argue that such a label should clip (I'd tend to clip it since its really a data annotation).  

This comes up for other artists as well like connection patches etc.
